quickwit/quickwit-janitor/src/garbage_collection.rs

```diff
@@ -186,30 +178,60 @@ pub async fn delete_splits_with_files(
     splits: Vec<SplitMetadata>,
     ctx_opt: Option<&ActorContext<GarbageCollector>>,
 ) -> anyhow::Result<Vec<FileEntry>, SplitDeletionError> {
-    let mut deleted_file_entries = Vec::new();
-    let mut deleted_split_ids = Vec::new();
-    let mut failed_split_ids_to_error = Vec::new();
+    let mut paths_to_splits = HashMap::with_capacity(splits.len());
+
+    for split in splits {
+        let file_entry = FileEntry::from(&split);
+        let split_filename = quickwit_common::split_file(split.split_id());
+        let split_path = Path::new(&split_filename);
 
-    let mut task_results_stream =
-        spawn_bulk_delete_tasks(MAX_CONCURRENT_STORAGE_REQUESTS, splits, storage, ctx_opt).await;
-    while let Some((split_ids, split_entries, delete_splits_res)) = task_results_stream.recv().await
-    {
-        if let Err(error) = delete_splits_res {
-            error!(error = ?error, index_id = ?index_id, split_ids = ?split_ids, "Failed to delete split.");
-            failed_split_ids_to_error.push((split_ids, error));
-        } else {
-            deleted_split_ids.extend(split_ids);
-            deleted_file_entries.extend(split_entries);
-        };
+        paths_to_splits.insert(
+            split_path.to_path_buf(),
+            (split.split_id().to_string(), file_entry),
+        );
     }
 
-    if !failed_split_ids_to_error.is_empty() {
-        error!(index_id = ?index_id, failed_split_ids_to_error = ?failed_split_ids_to_error, "Failed to delete splits.");
-        return Err(SplitDeletionError::StorageFailure(
-            failed_split_ids_to_error,
-        ));
+    let paths = paths_to_splits
+        .keys()
+        .map(|key| key.as_path())
+        .collect::<Vec<&Path>>();
+    let result = storage.bulk_delete(&paths).await;
+
+    if let Some(ctx) = ctx_opt {
+        ctx.record_progress();
     }
 
+    let mut deleted_split_ids = vec![];
+    let mut deleted_file_entries = vec![];
+
+    match result {
+        Ok(()) => {
+            for (split_id, entry) in paths_to_splits.into_values() {
+                deleted_split_ids.push(split_id);
+                deleted_file_entries.push(entry);
+            }
+        }
+        Err(state) => {
+            let split_ids = state
+                .failures
+                .keys()
+                .chain(state.unattempted.iter())
```

Comment:
Truncate to a few splits to avoid flooding the logs and collect into a string directly `Itertools::join`. Since we're truncating, display how many splits could not be deleted.

Reply:
I think truncating and joining a string could potentially be misleading or at the very least not be the most helpful way of displaying that information if we want to use the tracing values. I'd probably opt for just listing how many splits failed to be deleted by itself or we keep all of them displayed. 

If we do still want to display only the first few split ids, I would probably opt for displaying them as such:

```rust
            error!(
                error = ?bulk_delete_error.error, 
                index_id = ?index_id,
                num_split_ids = num_split_ids, 
                "Failed to {:?} and {} other splits.",
                truncated_splits_list, 
                num_split_ids
            );
```

The reason is that I don't think having a string value of `split_ids_sample = "phiu4h23423,uygeuyfg234,2g23uy4g2fgd"` is any more helpful or useful than displaying just a ` Vec<Path>` (which will also play nicer if we ever wanted to use tracing's ability to display logs in a JSON format or any of the other formatter they support)

Reply:
Most systems will truncate or reject oversized attribute values or spans. For instance, OTEL has limits for the number of attributes and length of values. New Relic limits value size to 4095 chars.

Logging potentially long values **here and in the rest of the codebase** is wrong.

Reply:
In that case, then I think we should go with the aforementioned `error!` log then.